File: services/client/src/protocol/protocol.go
package protocol

import (
	"encoding/binary"
	"io"

	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/safe_socket"
)

const (
	MsgBetBatch byte = 1
	MsgBatchAck byte = 2
	MsgDone     byte = 3
	MsgWinners  byte = 4
)

type Bet struct {
	FirstName string
	LastName  string
	Document  string
	Birthdate string
	Number    string
}

//Concatena agencyId + apuestas separadas por salto de linea
func Encode_bet_batch(agencyId string, bets []Bet) []byte {
	payload := agencyId + "|"
	for _, bet := range bets {
		payload += bet.FirstName + "," + bet.LastName + "," + bet.Document + "," +
			bet.Birthdate + "," + bet.Number + "\n"
	}
	return []byte(payload)
}

//Arma el Header de 5 bytes (tipo + longitud) y lo manda junto con el payload
func Write_message(w io.Writer, msgType byte, payload []byte) error {
	header := make([]byte, 5)
	header[0] = msgType
	binary.BigEndian.PutUint32(header[1:], uint32(len(payload)))
	return safe_socket.SendAll(w, append(header, payload...))
}

//Lee el Header de 5 bytes, extrae tipo y longitud, y lee el payload.
func Read_message(r io.Reader) (byte, []byte, error) {
	header, err := safe_socket.RecvAll(r, 5)
	if err != nil {
		return 0, nil, err
	}

	msgType := header[0]
	length := binary.BigEndian.Uint32(header[1:])
	payload := []byte{}	
	if length > 0{
		payload, err = safe_socket.RecvAll(r, int(length))
		if err != nil{
			return 0, nil, err
		}
	}

	return msgType, payload, nil
	
}

File: services/server/src/protocol/protocol.py
from lottery.bet import Bet

MSG_BET_BATCH = 1
MSG_BATCH_ACK = 2
MSG_DONE = 3
MSG_WINNERS = 4

_HEADER_SIZE = 5 # 1 byte para el tipo de mensaje + 4 bytes para el tamaño del payload

#Lee un mensaje, interpretando su Header y payload.
def read_message(safe_socket, socket):
    header = safe_socket.recv_all(socket, _HEADER_SIZE)
    msg_type = header[0]
    length = int.from_bytes(header[1:5], byteorder="big", signed=False)
    payload = safe_socket.recv_all(socket, length) if length > 0 else b""

    return msg_type, payload

#Construye y envia un mensaje siguiendo el formato del protocolo
def write_message(safe_socket, socket, msg_type: int, payload: bytes = b""):
    header = bytes([msg_type]) + len(payload).to_bytes(4, byteorder="big", signed=False)
    safe_socket.send_all(socket, header + payload)

#Convierte el payload de texto en una lista de objetos Bet
def decode_bets(agency_id: int, payload: bytes) -> list[Bet]:
    bets = []
    for line in payload.decode("utf-8").split("\n"):
        if not line:
            continue
        first_name, last_name, document, birthdate, number = line.split(",")
        bets.append(
            Bet(agency_id, first_name, last_name, int(document), birthdate, int(number))
        )
    return bets

#Convierte una lista de documentos ganadores en el payload de la respuesta
def encode_winners(documents: list[int]) -> bytes:
    return ",".join(str(doc) for doc in documents).encode("utf-8")